File: solutions/java/bowling/1/src/main/java/FrameContext.java
import java.util.List;

public final class FrameContext {

    private final List<Frame> frames;
    private final int frameIndex;

    public FrameContext(final List<Frame> frames, final int frameIndex) {
        this.frames = frames;
        this.frameIndex = frameIndex;
    }

    public Frame getCurrentFrame() {
        return this.frames.get(this.frameIndex);
    }

    public Frame getNextFrame() {
        int nextIndex = this.frameIndex + 1;
        return nextIndex < this.frames.size()
               ? this.frames.get(nextIndex)
               : null;
    }

    public Frame getFollowingFrame() {
        int followingIndex = this.frameIndex + 2;
        return followingIndex < this.frames.size()
               ? this.frames.get(followingIndex)
               : null;
    }

    public boolean isLastFrame() {
        return this.frameIndex == this.frames.size() - 1;
    }
}

File: solutions/ocaml/bowling/1/frame_context.ml
open Base

type t = {
  frames : Frame.frame list;
  frame_index : int;
}

let create (frames : Frame.frame list) (frame_index : int) : t =
  { frames; frame_index }

let get_current_frame (context : t) : Frame.frame =
  match List.nth context.frames context.frame_index with
  | Some frame -> frame
  | None -> failwith "Invalid frame index"

let get_next_frame (context : t) : Frame.frame option =
  List.nth context.frames (context.frame_index + 1)

let get_following_frame (context : t) : Frame.frame option =
  List.nth context.frames (context.frame_index + 2)

let is_last_frame (context : t) : bool =
  context.frame_index = List.length context.frames - 1
